docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```diff
@@ -0,0 +1,304 @@
+
+
+## Proposal
+
+A Hybrid design
+
+Evidence has the following simple interface:
+
+```go
+type Evidence interface {  //proposed
+  Height() int64                                     // height of the equivocation
+  Bytes() []byte                                     // bytes which comprise the evidence
+  Hash() []byte                                      // hash of the evidence
+  ValidateBasic() error
+  String() string
+}
+```
+
+We have two recognized evidence that complies with it
+
+```go
+type LightClientAttackEvidence struct {
+  ConflictingBlock *LightBlock
+  CommonHeight int64
+}
+```
+where the `Hash()` is the hash of the header and commonHeight
+
+```go
+type DuplicateVoteEvidence {
+  VoteA *Vote
+  VoteB *Vote
+}
+```
+where the `Hash()` is the hash of the two votes
+
+The first is generated in the light client, the second in consensus. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then  `Verify(ev Evidence) error` before adding it to the database
+
+
+`Verify()` does the following:
+
+- Use the hash to see if we already have this evidence in our committed database.
+
+- Use the height to check if the evidence hasn't expired.
+
+- If it hasn't expired use the height to find the block header and confirm that the time hasn't also expired.
+
+- Then proceed with switch statement for each of the two evidence:
+
+For `DuplicateVote`:
+
+- Check that height, round, type and validator address are the same
+
+- Check that the Block ID is different
+
+- Check the look up table for addresses to make sure there already isn't evidence against this validator
+
+- Fetch the validator set and confirm that the address is in the set
+
+- Check that the chain ID and signature is valid.
+
+For `LightClientAttack`
+
+- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header
+
+- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.
+
+  - If equivocation the validators that signed for the commits of both the trusted and signed header
+
+  - If lunatic the validators from the common val set that signed in the conflicting block
+
+  - If amnesia none (we can't know which validators are malicious). This also means that we don't
+  send amnesia evidence to the application
+
+- For each validator, check the look up table to make sure there already isn't evidence against this validator
+
+
+After verification we save the evidence with the following key `height/hash` and in the following struct
```

Comment:
When you say "save," are you talking about persisting the evidence to the aforementioned database? 

(Clearly I'm just missing this database from my mental model of the evidence flow.)

Reply:
....is this the same thing as the evidence pool?

Reply:
yes for both statements

Reply:
Wait, I'm thoroughly confused now - you said above that data is persisted to this db in case something happens to the evidence pool. Does the evidence pool persist data? Is it the only way that data is persisted?